xla/service/gpu/llvm_gpu_backend/amdgpu_backend.cc

```diff
@@ -117,60 +120,90 @@ std::vector<std::string> GetROCDLPaths(std::string gcn_arch_name,
+    static HsacoCache obj;
+    return obj;
+  }
+
+bool Find(const std::string& ir, const std::string& gfx, 
+        std::string *hash_str, std::vector<uint8_t> *hsaco, std::string *hsaco_path) {
+  std::lock_guard<std::mutex> lg(mutex_);
 
-static HsacoCache g_hsacoCache;  // NOLINT: static/global vars forbidden
+  llvm::SHA256 sha256;
+  sha256.update(llvm::StringRef(ir));
+  std::array<uint8_t, 32> lhash = sha256.final();
+  // C++ strict aliasing rules allow reinterpret casting to (const) char*.
+  absl::string_view hash_view(reinterpret_cast<const char*>(lhash.data()),
+                              lhash.size());
+  (void)tsl::Base64Encode(hash_view, hash_str);
+  // VLOG(0) << "Got hashview:" << hash_str;
+  *hash_str += "." + gfx;
 
-bool HsacoCache::Find(const std::string& ir, uint64_t& hash,
-                      const std::string& gfx, std::vector<uint8_t>& hsaco) {
-  std::lock_guard<std::mutex> lg(g_hsacoCache.m_mutex);
-  hash = std::hash<std::string>{}(ir);
   bool hit = false;
-  for (auto& x : g_hsacoCache.cache) {
-    if (x.hash != hash) continue;
-    if (x.gfx != gfx) continue;
-    if (x.ir != ir) continue;
-    hsaco = x.hsaco;
+  for (const auto& x : hsaco_cache_) {
+    if (!(x.hash_str == *hash_str && x.ir == ir)) continue;
+    *hsaco = x.hsaco;
     hit = true;
     break;
   }
-  g_hsacoCache.request_count++;
-  if (hit) g_hsacoCache.hit_count++;
-  if (!(g_hsacoCache.request_count % 50))
-    VLOG(1) << "HSACO cache: " << g_hsacoCache.request_count << " requests, "
-            << g_hsacoCache.hit_count << " hits";
+
+  *hsaco_path = hsaco_cache_dir_ + *hash_str + ".hsaco";
+  if (!hit && tsl::Env::Default()->FileExists(*hsaco_path).ok()) {
+      VLOG(1) << "Hsaco cache hit in file " << *hsaco_path;
+      std::ifstream hsaco_file(*hsaco_path, std::ios::binary | std::ios::ate);
+      std::ifstream::pos_type hsaco_file_size = hsaco_file.tellg();
+      *hsaco = std::vector<uint8_t>(hsaco_file_size);
+      hsaco_file.seekg(0, std::ios::beg);
+      hsaco_file.read(reinterpret_cast<char*>(hsaco->data()), hsaco_file_size);
+      hsaco_cache_.emplace_back(HsacoCacheEntry{*hash_str, ir, *hsaco});
+      hit = true;
+  }
+  request_count_++;
+  if (hit) hit_count_++;
+  VLOG(1) << "HSACO cache: " << request_count_ << " requests, "
+            << hit_count_ << " hits";
   return hit;
 }
 
-void HsacoCache::Add(const std::string& ir, uint64_t hash,
-                     const std::string& gfx,
+void Add(const std::string& ir, const std::string& hash_str,
                      const std::vector<uint8_t>& hsaco) {
-  std::lock_guard<std::mutex> lg(g_hsacoCache.m_mutex);
-  g_hsacoCache.cache.resize(g_hsacoCache.cache.size() + 1);
-  g_hsacoCache.cache.back().ir = ir;
-  g_hsacoCache.cache.back().hash = hash;
-  g_hsacoCache.cache.back().gfx = gfx;
-  g_hsacoCache.cache.back().hsaco = hsaco;
+  std::lock_guard<std::mutex> lg(mutex_);
+  hsaco_cache_.emplace_back(HsacoCacheEntry{hash_str, ir, hsaco});
```

Comment:
Don't you need to emit the file to the cache dir as well?

Reply:
cache dir path is already in the `str` from [HsacoCache::Find()](https://github.com/ROCm/xla/blob/103e8abf9e446f6dbc1249670f3b5cc418bfa1d0/xla/service/gpu/llvm_gpu_backend/amdgpu_backend.cc#L159-L161)